### DISCUSSION TITLE
Why can't I run the phases out of `nix develop`?

### DISCUSSION THREAD

**Discussion by Elias-Graf** [ipetkov/crane#697] [Q&A]
I'm trying to debug a flake (since it doesn't build with `nix build`, but does when running `cargo build` inside the dev shell), by executing the exact phases. But the devShell environment seems to be different from the build environment. For example, none of the phases seem to be populated:

```
[elias@perseus:~/Git/cosmic-files]$ echo $phases
buildPhase

[elias@perseus:~/Git/cosmic-files]$ echo $buildPhase
{ echo "------------------------------------------------------------"; echo " WARNING: the existence of this path is not guaranteed."; echo " It is an internal implementation detail for pkgs.mkShell."; echo "------------------------------------------------------------"; echo; # Record all build inputs as runtime dependencies export; } >> "$out"
```

I'm unsure if this is a general nix concept I'm not understanding, but I'm pretty sure that I could execute `eval "$buildPhase"` when building a simple "Hello World C" flake.

My flake is pretty much the one from the simple template I think, but here is the shell part if it's of any use:

```nix
        devShells.default = craneLib.devShell {
          # Inherit inputs from checks.
          checks = self.checks.${system};

          # Additional dev-shell environment variables can be set directly
          # MY_CUSTOM_DEVELOPMENT_VAR = "something else";

          # Extra inputs can be added here; cargo and rustc are provided by default.
          packages = [
            # pkgs.ripgrep
          ];
        };
```

Help and pointers greatly appreciated.

---
In case anyone is interested, what isn't working when running `nix build`: Pretty sure the issue is that the translation files do not get copied to the build environment due to the `cleanCargoSource`. But I would like to confirm that before just trying, so I can learn something. Thus, I'm trying to get the phases to work, which possible might get me the `src` if I run unpack(?).

**Comment by dpc** [comment]
Why would you want to do any building as a part of the dev shell? That makes no sense to me. I doubt craneLib.devShell (and underlying mkShell) even accept some of the arguments you're trying to check.

You probably want to create a derivation like in the examples, and build it with `nix build ...`, and maybe add the result of it to `packages` in your shell or something like that.

**Reply by Elias-Graf** [reply]
`nix build` does not build. I want to debug `nix build` by examining the environment. There is also a `runPhase` script available in the shell. You follow this: https://www.tweag.io/blog/2020-05-25-flakes/#writing-your-first-flake, you can see that it should be possible.

> $ nix develop
$ eval "$buildPhase"
$ ./hello
Hello World

**Comment by ipetkov** [comment]
Hi @Elias-Graf! You might need to explicitly call `nix develop .#mycrate` (or `nix develop .packages.x86_64-linux.mycrate`). By default `nix develop` will activate the default devshell (if defined), but that devshell only defines tools to bring in the environment, it isn't the actual package which has the build phases. 

BUT if you don't define a default devshell, `nix develop` will act as if you invoked it as `nix develop .packages.${currentSystem}.default` which might be leading to the confusion...

Hope this helps!
